jcl/src/java.base/share/classes/java/lang/System.java

```diff
@@ -176,18 +251,82 @@ static void afterClinitInitialization() {
+		/*[IF PLATFORM-mz31|PLATFORM-mz64]*/
+		try {
+			consoleDefaultCharset = Charset.forName(props.getProperty("ibm.system.encoding")); //$NON-NLS-1$
+		} catch (IllegalArgumentException e) {
+			// use the defaultCharset()
+		}
+		/*[ENDIF] PLATFORM-mz31|PLATFORM-mz64 */
+		Charset stdoutCharset = getCharset(props.getProperty("sun.stdout.encoding"), null); //$NON-NLS-1$
+		Charset stderrCharset = getCharset(props.getProperty("sun.stderr.encoding"), null); //$NON-NLS-1$
+		/*[ELSE]*/
+		Charset consoleCharset = Charset.defaultCharset();
+		String stdoutCharset = getCharsetName(props.getProperty("sun.stdout.encoding"), consoleCharset); //$NON-NLS-1$
+		String stderrCharset = getCharsetName(props.getProperty("sun.stderr.encoding"), consoleCharset); //$NON-NLS-1$
+		/*[ENDIF] JAVA_SPEC_VERSION >= 11 */
+		/*[ENDIF] Sidecar18-SE-OpenJ9 | JAVA_SPEC_VERSION >= 11 */
+
+		/*[IF (Sidecar18-SE-OpenJ9 | JAVA_SPEC_VERSION >= 11) & !(PLATFORM-mz31 | PLATFORM-mz64)]*/
+		if (stderrCharset != null) {
+			/*[IF JAVA_SPEC_VERSION >= 11]*/
+			hasSetErrEncoding = true;
+			setErr(new PrintStream(new BufferedOutputStream(new FileOutputStream(FileDescriptor.err)), true, stderrCharset));
+			/*[ELSE]*/
+			try {
+				setErr(new PrintStream(new BufferedOutputStream(new FileOutputStream(FileDescriptor.err)), true, stderrCharset));
+			} catch (UnsupportedEncodingException e) {
+				// not possible as the Charset name was validated first
+				throw new InternalError("For encoding " + stderrCharset, e); //$NON-NLS-1$
+			}
+			/*[ENDIF] JAVA_SPEC_VERSION >= 11 */
+		} else {
+			setErr(new PrintStream(new BufferedOutputStream(new FileOutputStream(FileDescriptor.err)), true));
+		}
+		if (stdoutCharset != null) {
+			/*[IF JAVA_SPEC_VERSION >= 11]*/
+			hasSetOutEncoding = true;
+			setOut(new PrintStream(new BufferedOutputStream(new FileOutputStream(FileDescriptor.out)), true, stdoutCharset));
+			/*[ELSE]*/
+			try {
+				setOut(new PrintStream(new BufferedOutputStream(new FileOutputStream(FileDescriptor.out)), true, stdoutCharset));
+			} catch (UnsupportedEncodingException e) {
+				// not possible as the Charset name was validated first
+				throw new InternalError("For encoding " + stdoutCharset, e); //$NON-NLS-1$
+			}
+			/*[ENDIF] JAVA_SPEC_VERSION >= 11 */
+		} else {
+			setOut(new PrintStream(new BufferedOutputStream(new FileOutputStream(FileDescriptor.out)), true));
+		}
 		/*[IF Sidecar19-SE_RAWPLUSJ9]*/
 		setIn(new BufferedInputStream(new FileInputStream(FileDescriptor.in)));
-		/*[ENDIF]*/
+		/*[ENDIF] Sidecar19-SE_RAWPLUSJ9 */
 		/*[ELSE]*/
 		/*[PR s66168] - ConsoleInputStream initialization may write to System.err */
 		/*[PR s73550, s74314] ConsolePrintStream incorrectly initialized */
+		/*[IF JAVA_SPEC_VERSION >= 11]*/
+		if (stderrCharset != null) {
+			hasSetErrEncoding = true;
+			setErr(com.ibm.jvm.io.ConsolePrintStream.localize(new BufferedOutputStream(new FileOutputStream(FileDescriptor.err)), true, stderrCharset));
```

Comment:
Since `Add ConsolePrintStream.localize with encoding` within JDK11 zOS confirmed that `com.ibm.jvm.io.ConsolePrintStream` is only required when `line.separator` is not `\n`.
I think a check can be added to limit the usage of `com.ibm.jvm.io.ConsolePrintStream` at the case that `line.separator` is not `\n` due to platform specific or command line settings.

Reply:
> Since Add ConsolePrintStream.localize with encoding within JDK11 zOS confirmed that com.ibm.jvm.io.ConsolePrintStream is only required when line.separator is not \n.

I don't think is true. ConsolePrintStream does a conversion when the file.encoding is different from the ibm.system.encoding.

Reply:
> ConsolePrintStream does a conversion when the file.encoding is different from the ibm.system.encoding.

As per `ConsolePrintStream` description:
```
 *     Wrapper class to perform NL to line.separator conversion on platforms
 *     where the line.separator != "\n". Intended for use on System.out and
 *     System.err .
```
If that's its sole purpose, there is no need to use `ConsolePrintStream` when current `line.separator == \n`.
There are conversions other than `\n` to `line.separator`, my question is, are they really needed? It might worth the effort to understand it and avoid the overhead of using `ConsolePrintStream` if possible.

Reply:
> There are conversions other than `\n` to `line.separator`, my question is, are they really needed? 

Somebody thought so, but it's not clear what the use cases are. We'll have to try and get that information from the IBM JCL team.

> It might worth the effort to understand it and avoid the overhead of using ConsolePrintStream if possible.

I expect so, but it seems unrelated to adding support for sun.stderr.encoding, sun.stdout.encoding and can be done separately. Considering IBM Java 8 uses it for all platforms, it seems very risky to consider removing it until all the use cases are understood. Removing it for Java 11 z/OS may be possible. Considering the ConsolePrintStream impl isn't part of OpenJ9, I don't think this is the right place to discuss it. Already you are copying comments from restricted materials to open source comments.

Reply:
> Removing it for Java 11 z/OS may be possible.

Yeah, that's what I was aiming for.
The comments referred I thought are from JCL team like other internal docs (RTC contents, etc.), in addition, this class is in IBM Java 8 all platforms.